### PR TITLE
Centralize virtualenv path construction

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Right now, we have virtualenv construction encoded in a few different places. Namely, it happens in both `gourgeist` and `virtualenv_layout.rs` -- _and_ `interpreter.rs` also encodes some knowledge about how they work, by way of reconstructing the `SysconfigPaths`.

Instead, `gourgeist` now returns the complete layout, enumerating all the directories it created. So, rather than returning a root directory, and re-creating all those paths in `uv-interpreter`, we pass the data directly back to it.

### CODE REVIEW

File: crates/gourgeist/src/bare.rs [hunk 1]
```diff
@@ -174,9 +156,19 @@ pub fn create_bare_venv(
...
     for (name, template) in ACTIVATE_TEMPLATES {
+        let relative_site_packages = if cfg!(unix) {
+            format!(
+                "../lib/{}{}.{}/site-packages",
+                interpreter.site_packages_python(),
+                interpreter.python_major(),
+                interpreter.python_minor(),
+            )
+        } else if cfg!(windows) {
+            "../Lib/site-packages".to_string()
```
Comment (charliermarsh) [inline crates/gourgeist/src/bare.rs:168]: I assume this is right? (This is the only behavior change in the PR -- I tried to do the above TODO.)

File: crates/uv-interpreter/src/python_environment.rs [hunk 2]
```diff
@@ -157,14 +152,35 @@ pub(crate) fn detect_virtual_env(layout: &VirtualenvLayout) -> Result<Option<Pat
             if !dot_venv.join("pyvenv.cfg").is_file() {
                 return Err(Error::MissingPyVenvCfg(dot_venv));
             }
-            let python = layout.python_executable(&dot_venv);
-            if !python.is_file() {
-                return Err(Error::BrokenVenv(dot_venv, python));
-            }
```
Comment (charliermarsh) [inline crates/uv-interpreter/src/python_environment.rs:163]: This check happens at the call sites.

File: crates/uv-interpreter/src/python_environment.rs [hunk 3]
```diff
@@ -157,14 +152,35 @@ pub(crate) fn detect_virtual_env(layout: &VirtualenvLayout) -> Result<Option<Pat
...
             debug!("Found a virtualenv named .venv at: {}", dot_venv.display());
             return Ok(Some(dot_venv));
         }
     }
 
     Ok(None)
 }
+
+/// Returns the path to the `python` executable inside a virtual environment.
+pub(crate) fn detect_python_executable(venv: impl AsRef<Path>) -> PathBuf {
```
Comment (charliermarsh) [inline crates/uv-interpreter/src/python_environment.rs:164]: We still encode a small amount of venv knowledge here, but I think it's ok. This method is actually _different_ than anything that happens elsewhere: we're trying to find the Python in an existing venv, which we may or may not have created.

File: crates/gourgeist/src/bare.rs [hunk 4]
```diff
@@ -290,17 +278,42 @@ pub fn create_bare_venv(
...
+            scripts: scripts.into_std_path_buf(),
+            platstdlib: platstdlib.into_std_path_buf(),
+            // Set `purelib` and `platlib` to the same value.
+            purelib: site_packages.clone().into_std_path_buf(),
+            platlib: site_packages.into_std_path_buf(),
+            // Inherited from the interpreter.
+            stdlib: interpreter.stdlib().to_path_buf(),
+            include: interpreter.include().to_path_buf(),
+            platinclude: interpreter.platinclude().to_path_buf(),
+            data: location.into_std_path_buf(),
```
Comment (charliermarsh) [inline crates/gourgeist/src/bare.rs:316]: Eventually, we should be creating these paths by using templatized paths from the base interpreter and filling them in. (This isn't correct on all platforms -- but this isn't a new problem, no behavior changes.)

File: crates/gourgeist/src/bare.rs [hunk 5]
```diff
@@ -106,16 +85,17 @@ pub fn create_bare_venv(
...
+
     let bin_name = if cfg!(unix) {
         "bin"
     } else if cfg!(windows) {
         "Scripts"
     } else {
         unimplemented!("Only Windows and Unix are supported")
     };
-    let bin_dir = location.join(bin_name);
+    let scripts = location.join(bin_name);
```
Comment (charliermarsh) [inline crates/gourgeist/src/bare.rs:98]: These are just renamed to match `sysconfig`.

File: crates/uv-interpreter/src/sysconfig.rs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+pub struct SysconfigPaths {
+    pub stdlib: PathBuf,
+    pub platstdlib: PathBuf,
+    pub purelib: PathBuf,
+    pub platlib: PathBuf,
+    pub include: PathBuf,
+    pub platinclude: PathBuf,
+    pub scripts: PathBuf,
+    pub data: PathBuf,
+}
```
Comment (charliermarsh) [inline crates/uv-interpreter/src/sysconfig.rs:18]: (This is moved out of `interpreter.rs`.)

File: crates/uv-interpreter/src/python_environment.rs [hunk 7]
```diff
@@ -157,14 +152,35 @@ pub(crate) fn detect_virtual_env(layout: &VirtualenvLayout) -> Result<Option<Pat
...
+    let venv = venv.as_ref();
+    if cfg!(windows) {
+        // Search for `python.exe` in the `Scripts` directory.
+        let executable = venv.join("Scripts").join("python.exe");
+        if executable.exists() {
+            return executable;
+        }
+
+        // Apparently, Python installed via msys2 on Windows _might_ produce a POSIX-like layout.
```
Comment (konstin) [inline crates/uv-interpreter/src/python_environment.rs:173]: Does this fix https://github.com/astral-sh/uv/issues/1251?

**Review by konstin — APPROVED** [review]

File: crates/gourgeist/src/bare.rs [hunk 8]
```diff
@@ -174,9 +156,19 @@ pub fn create_bare_venv(
     }
 
     // Add all the activate scripts for different shells
-    // TODO(konstin): RELATIVE_SITE_PACKAGES is currently only the unix path. We should ensure that all launchers work
-    // cross-platform.
     for (name, template) in ACTIVATE_TEMPLATES {
+        let relative_site_packages = if cfg!(unix) {
```
Comment (BurntSushi) [inline crates/gourgeist/src/bare.rs:160]: I'm wondering whether using conditional compilation for logic like this is the right thing to do. For example, are we sure that all `uv` binaries compiled for Windows will 100% of the time use the Windows oriented path below? Or is it possible for the Unix path (WSL?) to be used in a Windows context?

Comment (BurntSushi) [reply crates/gourgeist/src/bare.rs:160]: Same deal for use of conditional compilation elsewhere.

**Review by BurntSushi — APPROVED** [review]
Yesssssss! This makes a lot of sense. Thank you!

File: crates/gourgeist/src/bare.rs (hunk 8 above)
Comment (charliermarsh) [reply crates/gourgeist/src/bare.rs:160]: Yeah probably not, but I think it's ~the same as it is today. I have an idea for how to fix this properly (#2095).

Comment (konstin) [reply crates/gourgeist/src/bare.rs:160]: Unfortunately they don't (https://github.com/astral-sh/uv/issues/1251), but #2095 should solve either case

**Pull request merged**